src/transaction/flashback.c

```diff
@@ -0,0 +1,418 @@
+	       * so the current_time represents a time in the future than the log being read now
+	       * Since LOG_DUMMY_HA_SERVER_STATE is logged every second, start_time is set to 1 second before the current_time */
+
+	      summary_entry->start_time = current_time - 1;
+	      LSA_COPY (&summary_entry->start_lsa, &cur_log_rec_lsa);
+	    }
+	}
+
+      LOG_READ_ADD_ALIGN (thread_p, sizeof (*log_rec_header), &process_lsa, log_page_p);
+
+      switch (log_type)
+	{
+	case LOG_COMMIT:
+	  {
+	    LOG_REC_DONETIME *donetime;
+
+	    if (context->num_summary == FLASHBACK_MAX_SUMMARY)
+	      {
+		error = ER_FLASHBACK_TOO_MANY_SUMMARY;
+		goto exit;
+	      }
+
+	    LOG_READ_ADVANCE_WHEN_DOESNT_FIT (thread_p, sizeof (*donetime), &process_lsa, log_page_p);
+
+	    donetime = (LOG_REC_DONETIME *) (log_page_p->area + process_lsa.offset);
+
+	    /* create summary info entry, and register it to the summary list */
+	    if ((error = flashback_create_summary_entry (thread_p, &summary_entry)) != NO_ERROR)
+	      {
+		goto exit;
+	      }
+
+	    summary_entry->trid = trid;
+	    LSA_COPY (&summary_entry->end_lsa, &cur_log_rec_lsa);
+	    summary_entry->end_time = donetime->at_time;
+
+        // *INDENT-OFF*
+	context->summary_list.insert (std::make_pair (trid, summary_entry));
+        // *INDENT-ON*
+	    current_time = donetime->at_time;
+	    context->num_summary++;
+	    break;
+	  }
+	case LOG_DUMMY_HA_SERVER_STATE:
+	  {
+	    LOG_REC_HA_SERVER_STATE *ha_dummy;
+
+	    LOG_READ_ADVANCE_WHEN_DOESNT_FIT (thread_p, sizeof (*ha_dummy), &process_lsa, log_page_p);
+
+	    ha_dummy = (LOG_REC_HA_SERVER_STATE *) (log_page_p->area + process_lsa.offset);
+
+	    current_time = ha_dummy->at_time;
+
+	    break;
+	  }
+	case LOG_SUPPLEMENTAL_INFO:
+	  {
+	    LOG_REC_SUPPLEMENT *supplement;
+	    int supplement_length;
+	    SUPPLEMENT_REC_TYPE rec_type;
+	    RECDES supp_recdes = RECDES_INITIALIZER;
+
+	    OID classoid;
+
+	    LOG_READ_ADVANCE_WHEN_DOESNT_FIT (thread_p, sizeof (*supplement), &process_lsa, log_page_p);
+
+	    supplement = (LOG_REC_SUPPLEMENT *) (log_page_p->area + process_lsa.offset);
+	    supplement_length = supplement->length;
+	    rec_type = supplement->rec_type;
+
+	    LOG_READ_ADD_ALIGN (thread_p, sizeof (*supplement), &process_lsa, log_page_p);
+
+	    if (cdc_get_undo_record (thread_p, log_page_p, cur_log_rec_lsa, &supp_recdes) != S_SUCCESS)
+	      {
+		error = ER_FAILED;
+		goto exit;
+	      }
+
+	    supplement_length = sizeof (supp_recdes.type) + supp_recdes.length;
+	    if (supplement_length > supplement_alloc_length)
+	      {
+		char *tmp = (char *) db_private_realloc (thread_p, supplement_data, supplement_length);
+		if (tmp == NULL)
+		  {
+		    er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, supplement_length);
+		    error = ER_OUT_OF_VIRTUAL_MEMORY;
+		    goto exit;
+		  }
+		else
+		  {
+		    supplement_data = tmp;
+		    supplement_alloc_length = supplement_length;
+		  }
```

Comment:
Why don't you assign directly to `supplement_data`?
```suggestion
		char *supplement_data = (char *) db_private_realloc (thread_p, supplement_data, supplement_length);
		if (supplement_data == NULL)
		  {
		    er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, supplement_length);
		    error = ER_OUT_OF_VIRTUAL_MEMORY;
		    goto exit;
		  }
		supplement_alloc_length = supplement_length;
```

Reply:
if realloc failed, then the pointer for existing memory area can be lost. 
memory is still allocated for supplement_data, but pointer (supplement_data) is assigned to NULL by return value of realloc(). 
So, pointer `tmp` is to avoid above situation.

Reply:
oh, you're right. I misunderstood.